bin/lib/invalidUsernames.js

```diff
@@ -0,0 +1,160 @@
+const fs = require('fs-extra')
+const Handlebars = require('handlebars')
+const path = require('path')
+const { URL } = require('url')
+const util = require('util')
+
+const { loadConfig } = require('./common')
+const { isValidUsername } = require('../../lib/common/user-utils')
+const blacklistService = require('../../lib/services/blacklist-service')
+const { initConfigDir, initTemplateDirs } = require('../../lib/server-config')
+const { fromServerConfig } = require('../../lib/models/oidc-manager')
+
+const AccountManager = require('../../lib/models/account-manager')
+const EmailService = require('../../lib/services/email-service')
+const LDP = require('../../lib/ldp')
+const SolidHost = require('../../lib/models/solid-host')
+
+const fileExists = util.promisify(fs.exists)
+const fileRename = util.promisify(fs.rename)
+
+module.exports = function (program) {
+  program
+    .command('invalidusernames')
+    .option('--notify', 'Will notify users with usernames that are invalid')
+    .option('--delete', 'Will delete users with usernames that are invalid')
+    .description('Manage usernames that are invalid')
+    .action(async (options) => {
+      const config = await loadConfig(program, options)
+      if (!config.multiuser) {
+        return console.error('You are running a single user server, no need to check for invalid usernames')
+      }
+
+      const invalidUsernames = await getInvalidUsernames(config)
+      const host = SolidHost.from({ port: config.port, serverUri: config.serverUri })
+      const accountManager = getAccountManager(config, host)
+
+      if (options.notify) {
+        return notifyUsers(invalidUsernames, accountManager, config)
+      }
+
+      if (options.delete) {
+        return deleteUsers(invalidUsernames, accountManager, config, host)
+      }
+
+      listUsernames(listUsernames)
+    })
+}
+
+async function createNewIndexFile (username, accountManager, invalidUsernameTemplate, dateOfRemoval, supportEmail, fileOptions) {
```

Comment:
Personally I don't like functions with so many arguments. I consider them a code smell.
Like the function trying to do too much.
It helps to have smaller functions, which are then called in order.

Reply:
I don't like long signatures either, but in this case it's not really reusable functions, but I wanted more readable code by splitting it up into "logical" chunks of code (in what I hope is good named functions).

But yeah, if we wanted to reuse this functionality, I should probably restructure it a bit to decrease the length of the signatures.

Reply:
Well, if you are struggling with writing tests, this should be argument enough to break the function up, shouldn't it?

Reply:
If I knew how to break up this code into reusable, testable functions, I would do it ;)

Reply:
Okay, I can write the high-level steps I see. Implementation goes into the functions then …

```js
async function createNewIndexFile (/* arguments */) {
  if (isBackupNeeded(/* arguments */)) {
    await Promise.all([
      backupIndexHtml(/* arguments */)
      backupAcl(/* arguments */)
    ])
    await Promise.all([
      generateNewIndexHtml(/* arguments */)
      generateNewAcl(/* arguments */)
    ])
    logCompletion(/* arguments */)
  }
}
```
or maybe group the backup + creation into a single function (like an transaction).
You may want to timestamp the `index.backup.html` also.

Reply:
Plus, looking at the function name, `createNewIndexFile` is misleading, since it does more than it says …
so its caller would need to call more functions.